File: ios/Approach/Sources/StatisticsLibrary/Protocols/PercentageStatistic.swift
import Charts
import StringsLibrary

public protocol PercentageStatistic: Statistic {
	static var numeratorTitle: String { get }
	static var denominatorTitle: String { get }
	static var includeNumeratorInFormattedValue: Bool { get }

	var numerator: Int { get set }
	var denominator: Int { get set }
	var percentage: Double { get }
}

extension PercentageStatistic {
	public static var supportsAggregation: Bool {
		true
	}

	public static var supportsWidgets: Bool {
		true
	}

	public var percentage: Double {
		denominator > 0 ? Double(numerator) / Double(denominator) : 0
	}

	public mutating func aggregate(with: Statistic) {
		guard let with = with as? Self else { return }
		self.numerator += with.numerator
		self.denominator += with.denominator
	}

	public var formattedValue: String {
		format(percentage: percentage)
	}

	public var formattedValueDescription: String? {
		Self.includeNumeratorInFormattedValue && numerator > 0
			? "\(numerator)/\(denominator)"
			: nil
	}

	public var isEmpty: Bool {
		denominator == 0
	}
}

File: android/core/statistics/src/main/java/ca/josephroque/bowlingcompanion/core/statistics/interfaces/PercentageStatistic.kt
package ca.josephroque.bowlingcompanion.core.statistics.interfaces

import ca.josephroque.bowlingcompanion.core.statistics.Statistic
import ca.josephroque.bowlingcompanion.core.statistics.utils.formatAsPercentage

interface PercentageStatistic: Statistic {
	val numeratorTitleResourceId: Int
	val denominatorTitleResourceId: Int
	val includeNumeratorInFormattedValue: Boolean

	var numerator: Int
	var denominator: Int

	override val supportsAggregation: Boolean
		get() = true

	override val supportsWidgets: Boolean
		get() = true

	val percentage: Double
		get() = if (denominator > 0) numerator.toDouble() / denominator.toDouble() else 0.0

	override val formattedValue: String
		get() = percentage.formatAsPercentage

	override val formattedValueDescription: String?
		get() = if (includeNumeratorInFormattedValue && numerator > 0) {
			String.format("%d/%d", numerator, denominator)
		} else {
			null
		}

	override val isEmpty: Boolean
		get() = denominator == 0

	override fun aggregateWithStatistic(statistic: Statistic) {
		if (statistic is PercentageStatistic) {
			numerator += statistic.numerator
			denominator += statistic.denominator
		}
	}
}